pyccel/parser/semantic.py

```diff
@@ -3608,13 +3621,35 @@ def _visit_ClassDef(self, expr):
         for i in methods:
             self._visit_FunctionDef(i)
 
+        if not any(method.name == '__del__' for method in methods):
+            argument = FunctionDefArgument(Variable(cls.name, 'self', cls_base = cls))
+            scope = self.create_new_function_scope('__del__')
+            del_method = FunctionDef('__del__', [argument], (), [Pass()], cls_name=cls.name, scope=scope)
+            self.exit_function_scope()
+            self.insert_function(del_method)
+            cls.add_new_method(del_method)
+
+        for method in cls.methods:
+            if method.name == '__del__':
+                self._current_function = method.name
+                attribute = [attr for attr in cls.attributes if not attr.on_stack]
+                if attribute:
+                    self._allocs[-1].extend(attribute)
+                    is_freed = next(is_freed for is_freed in cls.attributes if is_freed.name.startswith("is_freed"))
+                    method.body.insert2body(*self._garbage_collector(method.body))
+                    condition = If(IfSection(PyccelNot(is_freed),
+                                    [method.body]+[Assign(is_freed, LiteralTrue())]))
+                    method.body = [condition]
+                self._current_function = None
+                break
+
         self.exit_class_scope()
 
         return EmptyNode()
 
     def _visit_Del(self, expr):
 
-        ls = [self._visit(i) for i in expr.variables]
+        ls = [Deallocate(self._visit(i)) for i in expr.variables]
         return Del(ls)
```

Comment:
Why not return a `CodeBlock`? That way you can completely avoid the need for `_print_Del`

Reply:
In the Python printer it is important to keep `Del` objects as these should be printed while `Deallocate` objects shouldn't be